snapd-glib/snapd-notices-monitor.c

```diff
@@ -0,0 +1,166 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "snapd-notices-monitor.h"
+
+struct _SnapdNoticesMonitor {
+    GObject parent_instance;
+
+    SnapdClient *client;
+    GCancellable *cancellable;
+    GDateTime *last_date_time;
+    gdouble last_date_time_seconds;
+};
+
+G_DEFINE_TYPE(SnapdNoticesMonitor, snapd_notices_monitor, G_TYPE_OBJECT)
+
+static void begin_monitor(SnapdNoticesMonitor *self);
+
+static void monitor_cb(SnapdClient* source,
+                       GAsyncResult* res,
+                       SnapdNoticesMonitor *self) {
+
+    g_autoptr(GPtrArray) notices = NULL;
+    g_autoptr(GError) error = NULL;
+
+    notices = snapd_client_get_notices_finish(source, res, &error);
+
+    if (error != NULL) {
+        g_clear_object(&self->cancellable);
+        g_clear_pointer(&self->last_date_time, g_date_time_unref);
+        self->last_date_time_seconds = -1.0;
+        g_signal_emit_by_name(self, "error-event", error);
+        return;
+    }
+
+    if (g_cancellable_is_cancelled(self->cancellable)) {
+        g_clear_object(&self->cancellable);
+        g_clear_pointer(&self->last_date_time, g_date_time_unref);
+        self->last_date_time_seconds = -1.0;
+        return;
+    }
+
+    if ((error == NULL) && (notices != NULL)) {
+        for (int i=0; i < notices->len; i++) {
+            g_autoptr(SnapdNotice) notice = g_object_ref(notices->pdata[i]);
+            g_autoptr(GDateTime) last_occurred = (GDateTime *)snapd_notice_get_last_occurred(notice);
+            gdouble last_occurred_seconds = snapd_notice_get_last_occurred_seconds(notice);
+
+            if ((self->last_date_time == NULL) ||
+                (g_date_time_compare(self->last_date_time, last_occurred) == -1) ||
+                ((g_date_time_compare(self->last_date_time, last_occurred) == 0) && (self->last_date_time_seconds < last_occurred_seconds))) {
+                    g_clear_pointer (&self->last_date_time, g_date_time_unref);
+                    self->last_date_time = g_date_time_ref(last_occurred);
+                    self->last_date_time_seconds = snapd_notice_get_last_occurred_seconds(notice);
+                }
+            g_signal_emit_by_name(self, "notice-event", notice);
+        }
+    }
+    begin_monitor(self); // continue waiting for new notifications
+}
+
+static void begin_monitor(SnapdNoticesMonitor *self) {
+    snapd_client_set_notices_filter_by_date_seconds (self->client, self->last_date_time_seconds);
+    snapd_client_get_notices_async(self->client,
+                                   self->last_date_time,
+                                   200000000000000, // "infinity"
```

Comment:
G_INT64_MAX?

Reply:
Already tried with G_MAXINT64, but it seems that there is a time limit in snapd and returns `invalid timeout: time: invalid duration "9223372036854775807us"`. I did some tests, and the maximum value seems to be around 9223300000000000.